stonesoup/regulariser/particle.py

```diff
@@ -0,0 +1,108 @@
+    random number assigned to each particle from a uniform random distribution, otherwise they
+    remain the same. Further details on the implementation are given in [2]_.
+
+    References
+    ----------
+    .. [1] Robert, Christian P. & Casella, George, Monte Carlo Statistical Methods, Springer, 1999.
+
+    .. [2] Ristic, Branco & Arulampalam, Sanjeev & Gordon, Neil, Beyond the Kalman Filter:
+        Particle Filters for Target Tracking Applications, Artech House, 2004. """
+
+    def regularise(self, prior, posterior, detections):
+        """Regularise the particles
+
+        Parameters
+        ----------
+        prior : :class:`~.ParticleState` type or list of :class:`~.Particle`
+            prior particle distribution
+        posterior : :class:`~.ParticleState` type or list of :class:`~.Particle`
+            posterior particle distribution
+        detections : set of :class:`~.Detection`
+            set of detections containing clutter,
+            true detections or both
+
+        Returns
+        -------
+        particle state: :class:`~.ParticleState`
+           The particle state after regularisation
+        """
+
+        if not isinstance(posterior, ParticleState):
+            posterior = copy.copy(posterior)
+            posterior = ParticleState(None, particle_list=posterior)
+
+        if not isinstance(prior, ParticleState):
+            prior = copy.copy(prior)
+            prior = ParticleState(None, particle_list=prior)
+
+        regularised_particles = copy.copy(posterior)
+        moved_particles = copy.copy(posterior)
+
+        if detections is not None:
+            ndim = prior.state_vector.shape[0]
+            nparticles = len(posterior)
+
+            measurement_model = next(iter(detections)).measurement_model
+
+            # calculate the optimal bandwidth for the Gaussian kernel
+            hopt = (4/(ndim+2))**(1/(ndim+4))*nparticles**(-1/(ndim+4))
+            covar_est = posterior.covar
+
+            # move particles
+            moved_particles.state_vector = moved_particles.state_vector + \
+                hopt * cholesky_eps(covar_est) @ np.random.randn(ndim, nparticles)
+
+            # Evaluate likelihoods
+            part_diff = np.array(moved_particles.state_vector - prior.state_vector).T
+            part_diff_mean = np.mean(part_diff, axis=0)
+            move_likelihood = multivariate_normal.pdf(part_diff,
+                                                      mean=part_diff_mean,
+                                                      cov=np.array(covar_est))
+            post_part_diff = np.array(posterior.state_vector - prior.state_vector).T
+            post_part_diff_mean = np.mean(post_part_diff, axis=0)
+            post_likelihood = multivariate_normal.pdf(post_part_diff,
+                                                      mean=post_part_diff_mean,
+                                                      cov=covar_est)
```

Comment:
The underlying `pdf` method does `x - mean` after converting everything to standard float array, but by calculating difference prior to calling `pdf` and leaving the mean as `0`, this allows the custom typing in `StateVectors` to be used e.g. handle angle wrapping.

I also swapped `mean` to `average` as it turns out `mean` override on `StateVectors` is completely broken at the moment, which I'll fix in another PR.

```suggestion
            # Evaluate likelihoods
            part_diff = moved_particles.state_vector - prior.state_vector
            part_diff_mean = np.average(part_diff, axis=1)
            move_likelihood = multivariate_normal.pdf((part_diff - part_diff_mean).T,
                                                      cov=covar_est)
            post_part_diff = posterior.state_vector - prior.state_vector
            post_part_diff_mean = np.average(post_part_diff, axis=1)
            post_likelihood = multivariate_normal.pdf((post_part_diff - post_part_diff_mean).T,
                                                      cov=covar_est)
```